pact/dao/dao.pact

```diff
@@ -0,0 +1,279 @@
+      {"guard":=guard, "active":=active}
+      (enforce-guard guard)
+      (enforce active "Ambassador acct is disabled")))
+  (defun view-ambassadors ()
+    (with-capability (INTERNAL)
+      (map (row-with-key ambassadors) (keys ambassadors))))
+
+  (defun register-ambassador:bool (guardian:string acct:string guard:guard)
+    (with-capability (GUARDIAN guardian)
+      (insert ambassadors acct
+              {"guard":guard
+              ,"active":true
+              ,"voted-to-freeze":(time "2000-01-01T12:00:00Z")})
+      (with-capability (INTERNAL)
+        (adjust-ambassador-count 1)))
+      true)
+  (defun deactivate-ambassador:bool (guardian:string ambassador:string)
+    (with-capability (GUARDIAN guardian)
+      (let ((lst-deactivate (at 'last-ambassador-deactivation (read state DAO_STATE_KEY))))
+        (enforce (> (now) (add-time lst-deactivate DEACTIVATE_COOLDOWN)) "Deactivate Cooldown Failure")
+        (update state DAO_STATE_KEY {"last-ambassador-deactivation":(now)})
+        (update ambassadors ambassador {"active":false}))
+      (with-capability (INTERNAL)
+        (adjust-ambassador-count -1)))
+      true)
+  (defun reactivate-ambassador:bool (guardian:string ambassador:string)
+    (with-capability (GUARDIAN guardian)
+      (update ambassadors ambassador {"active":true})
+      (with-capability (INTERNAL)
+        (adjust-ambassador-count 1)))
+      true)
+
+  ; ----
+  ; Freeze the DAO
+  (defun vote-to-freeze:bool (ambassador:string)
+    (with-capability (AMBASSADOR ambassador)
+      (update ambassadors ambassador {"voted-to-freeze":(now)}))
+    true)
+
+  (defun freeze:string (ambassador:string)
+    (with-capability (AMBASSADOR ambassador)
+    (let* ((live-ambs
+              (map (at "voted-to-freeze")
+              (select ambassadors ["voted-to-freeze"] (where 'active (= true)))))
+           (ambs-cnt (length live-ambs))
+           (ambs-voted-to-freeze (length (filter (btwn-incl (add-time (now) (- APPROVAL_COOLDOWN)) (now)) live-ambs)))
+           (dao-frz-ts (add-time (now) FREEZE_TIMEOUT)))
+      (enforce (> (* 2 ambs-voted-to-freeze) ambs-cnt)
+        (format "Majority vote failed: {} of {}" [ambs-voted-to-freeze ambs-cnt]))
+      (is-dao-frozen) ; so it can't get called twice
+      ; is this correct?
+      (update state DAO_STATE_KEY
+        {"dao-frozen-until": dao-frz-ts})
+      (format "DAO now frozen until {}" [dao-frz-ts]))))
+
+
+  ; ----
+  ; Upgrade the DAO
+  (defcap GOVERNANCE ()
+    (if (try false (require-capability (INTERNAL)))
+      true
+      (let ((x (is-dao-frozen)); if it's frozen, bail
+            (res (check-hash-approval (tx-hash))))
+      res)))
+
+  (defun check-hash-approval:bool (hsh:string)
+    (with-read state DAO_STATE_KEY
+      {'proposed-upgrade-time:=prp-time
+      ,'guardian-count:=grd-cnt}
+      (enforce (>= (now) (add-time prp-time APPROVAL_COOLDOWN))
+        (format "Proposal still in cooldown" [(add-time prp-time APPROVAL_COOLDOWN)]))
```

Comment:
No `{}`s in format string. But also, in general consider "static" errors ie just `(enforce ... "Proposal still in cooldown")` as they are easier to verify, shorter etc.